packages/clay-modal/src/ClayModal.tsx

```diff
@@ -0,0 +1,206 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import Button from '@clayui/button';
+import classNames from 'classnames';
+import Icon from '@clayui/icon';
+import React, {FunctionComponent, useEffect, useRef, useState} from 'react';
+import {createPortal} from 'react-dom';
+import {Footer, FooterContainer, FooterContainerProps} from './Footer';
+import {useModal} from './Hook';
+
+type Size = 'full-screen' | 'lg' | 'sm';
+
+type Status = 'danger' | 'info' | 'success' | 'warning';
+
+interface Props extends React.HTMLAttributes<HTMLDivElement> {
+	/**
+	 * Callback called to close the modal.
+	 */
+	onClose: () => void;
+
+	/**
+	 * Renders the footer with the FooterContainer component utilities.
+	 */
+	renderFooter?: React.ReactElement;
+
+	/**
+	 * The size of element modal.
+	 */
+	size?: Size;
+
+	/**
+	 * The path to the SVG spritemap file containing the icons.
+	 */
+	spritemap?: string;
+
+	/**
+	 * Status messages.
+	 */
+	status?: Status;
+
+	/**
+	 * Title of the modal.
+	 */
+	title?: string;
+
+	/**
+	 * Url to place an iframe in the body of the modal.
+	 */
+	url?: string;
+
+	/**
+	 * Flag to indicate if the modal is open or closed.
+	 */
+	visible: boolean;
+}
+
+const ICON_MAP = {
+	danger: 'exclamation-full',
+	info: 'info-circle',
+	success: 'check-circle',
+	warning: 'question-circle-full',
+};
+
+const Portal = ({children}: {children: React.ReactNode}) => {
+	const portalRef = useRef(document.createElement('div'));
+	const elToMountTo = document && document.body;
+
+	useEffect(() => {
+		elToMountTo.appendChild(portalRef.current);
+		return () => {
+			elToMountTo.removeChild(portalRef.current);
+		};
+	}, [elToMountTo]);
+
+	return createPortal(children, portalRef.current);
+};
+
+const delay = (fn: Function) => {
+	const callbackId = setTimeout(() => {
+		fn();
+		clearTimeout(callbackId);
+	}, 100);
+};
+
+const ClayModal: FunctionComponent<Props> & {
+	FooterContainer: FunctionComponent<FooterContainerProps>;
+} = ({
+	children,
+	className,
+	onClose = () => {},
+	renderFooter,
+	size,
+	spritemap,
+	status,
+	title,
+	url,
+	visible = false,
+	...otherProps
+}) => {
+	/**
+	 * It is set to keep track of modal visibility and ensure that classes
+	 * are removed when necessary.
+	 */
+	const [visibleModal, setVisibleModal] = useState<boolean>(visible);
+
+	const [visibleClassShow, setVisibleClassShow] = useState<boolean>(false);
+
+	const modalDialogElementRef = useRef<HTMLDivElement | null>(null);
+
+	/**
+	 * A hook that takes care of controlling click, keyup and keydown events
+	 * respectively close the modal after a click on the overlay, close the
+	 * modal by pressing the ESC key and control the focus within the Modal.
+	 */
+	useModal(modalDialogElementRef, visible, onClose);
```

Comment:
Since this hook is concerned with interactions, could we name it `useUserInteractions` or something like that? `useModal` is a big ambiguous to me.

Reply:
Yeah, I did not have the creativity to name that hook 😅. `useUserInteractions` sounds good.